### ISSUE TITLE
Cloud Foundry Service Registry (Eureka) fails Blue-Green deployment & route paths

### ISSUE THREAD

**Issue by odedia** [pivotal-cf/spring-cloud-services-connector#23]
Hi,
I face two issues related to deploying microservices to cloud foundry with Spring Cloud Services enabled.
First, it appears that if I deploy an app with CF route path, if would simply fail to register. I deploy an app like this:

`cf push -p target/date-services-1.0.jar --hostname env5-date-services --route-path /v1 -f manifest.yml date-services-v1
`
The manifest looks like this:

```
applications:
- name: date-services-v1
  memory: 1024M
  instances: 1
  timeout: 180
  env:
    JBP_CONFIG_OPEN_JDK_MEMORY_CALCULATOR: "[memory_sizes: {metaspace: 180m}, memory_heuristics: {metaspace:  25}]"
  services:
  - rabbitmq
  - config-server
  - discovery
  - circuit-breaker
  - mysql

```

The application gets registered by Eureka, but it cannot be found. If I open the link from the Eureka dashboard, it is translated as something like:

http://site.com/v1:80/info
http://site.com/v1:80/date

This URL is simply inaccessible. Other apps try to call the date-services microservice and fail with "404 not found". If I remove the route path - from the deployment - everything is great.

The other problem is with blue-green deployments.

If I deploy the same jar to a new app name, like date-services-GREEN, again I am failing in Eureka:

```
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT 2017-03-23 07:21:16.225 ERROR [date-services-sa-v1,,,] 20 --- [tbeatExecutor-0] com.netflix.discovery.DiscoveryClient : DiscoveryClient_date-services-SA-V1/pcfdev01-date-services-sa-v1-GREEN.apps.cf.site.com:84e48618-3b11-46e4-7b3e-f638788786ea - was unable to send heartbeat!
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient.register(DiscoveryClient.java:815)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT 2017-03-23 07:21:16.225 WARN [date-services-sa-v1,,,] 20 --- [tbeatExecutor-0] com.netflix.discovery.DiscoveryClient : DiscoveryClient_date-services-SA-V1/pcfdev01-date-services-sa-v1-GREEN.apps.cf.site.com:84e48618-3b11-46e4-7b3e-f638788786ea - registration failed Cannot execute request on any known server
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT com.netflix.discovery.shared.transport.TransportException: Cannot execute request on any known server
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:111)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient.renew(DiscoveryClient.java:837)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient$HeartbeatThread.run(DiscoveryClient.java:1396)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$1.execute(EurekaHttpClientDecorator.java:59)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.lang.Thread.run(Thread.java:745)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT com.netflix.discovery.shared.transport.TransportException: Cannot execute request on any known server
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:111)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient.register(DiscoveryClient.java:815)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient.renew(DiscoveryClient.java:837)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient$HeartbeatThread.run(DiscoveryClient.java:1396)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT at java.lang.Thread.run(Thread.java:745)
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT 2017-03-23 07:21:16.306 INFO [date-services-sa-v1,,,] 20 --- [nfoReplicator-0] com.netflix.discovery.DiscoveryClient : DiscoveryClient_date-services-SA-V1/pcfdev01-date-services-sa-v1-GREEN.apps.cf.site.com:84e48618-3b11-46e4-7b3e-f638788786ea: registering service...
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT com.netflix.discovery.shared.transport.TransportException: Cannot execute request on any known server
Thu Mar 23 2017 09:21:16 GMT+0200 (IST) [APP] OUT 2017-03-23 07:21:16.331 WARN [date-services-sa-v1,,,] 20 --- [nfoReplicator-0] c.n.d.s.t.d.RetryableEurekaHttpClient : Request execution failure with status code 403; retrying on another server if available

```

Eventually these errors go away, however after full blue-green the registered route in Eureka still points to GREEN and not the original route. Which is ofcourse inaccessible.

Here is the commands I am running:

```
cf push -p artifact.jar --hostname ${ENV_NAME}-${HOST_NAME}-${VERSION_NUMBER}-GREEN  -f manifest.yml ${TARGET_JOB_NAME}-GREEN

cf map-route ${TARGET_JOB_NAME}-GREEN ${DOMAIN} --hostname ${ENV_NAME}-${HOST_NAME}-${VERSION_NUMBER}

cf delete-route ${DOMAIN} -f --hostname ${ENV_NAME}-${HOST_NAME}-${VERSION_NUMBER}-GREEN

cf rename ${TARGET_JOB_NAME}-GREEN ${TARGET_JOB_NAME}
```

The final URL that is registered in Eureka is:

http://pcfdev01-date-services-sa-v1-green.cf.site.com/info

With the "green" in the URL, it simply doesn't work. If I remove the "green" from the URL, it works.

I am using Camden.SR6 and Spring Cloud Services 1.4.1.RELEASE. I Believe the runtime in the CF environment is 1.7.

Thanks.

**Comment by odedia** [comment]
Further information:

I performed the blue-green deployment manually "step by step" to identify the source of issue.

Even at the first green deployment I am failing.

I already have the "old" app in CF. I uploaded the "old" app with:

cf push -p date-service-sa-v1.jar  --hostname env1-date-v1 -f manifest.yml date-v1

Everything was fine.

Then I deployed the green app:

cf push -p date-service-sa-v1.jar  --hostname env1-date-v1-GREEN -f manifest.yml date-v1-GREEN

Already at this stage, it fails to register with Eureka:

```
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT 2017-03-23 08:43:15.391 INFO [date-service-sa-v1,,,] 19 --- [nfoReplicator-0] com.netflix.discovery.DiscoveryClient : DiscoveryClient_date-SERVICE-SA-V1/env1-date-v1-GREEN.apps.cf.site.com:91dfe266-422d-4f5a-69ec-9caf12889a1d: registering service...
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT 2017-03-23 08:43:15.459 INFO [date-service-sa-v1,,,] 19 --- [ main] o.s.i.endpoint.EventDrivenConsumer : Adding {logging-channel-adapter:_org.springframework.integration.errorLogger} as a subscriber to the 'errorChannel' channel
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT 2017-03-23 08:43:15.459 INFO [date-service-sa-v1,,,] 19 --- [ main] o.s.i.endpoint.EventDrivenConsumer : started _org.springframework.integration.errorLogger
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT 2017-03-23 08:43:15.459 INFO [date-service-sa-v1,,,] 19 --- [ main] o.s.i.channel.PublishSubscribeChannel : Channel 'date-service-sa-v1:cloud:0.errorChannel' has 1 subscriber(s).
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT 2017-03-23 08:43:15.462 WARN [date-service-sa-v1,,,] 19 --- [nfoReplicator-0] c.n.d.s.t.d.RetryableEurekaHttpClient : Request execution failure with status code 403; retrying on another server if available
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT 2017-03-23 08:43:15.467 WARN [date-service-sa-v1,,,] 19 --- [nfoReplicator-0] com.netflix.discovery.DiscoveryClient : DiscoveryClient_date-SERVICE-SA-V1/env1-date-v1-GREEN.apps.cf.site.com:91dfe266-422d-4f5a-69ec-9caf12889a1d - registration failed Cannot execute request on any known server
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:111)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$1.execute(EurekaHttpClientDecorator.java:59)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.InstanceInfoReplicator$1.run(InstanceInfoReplicator.java:88)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.lang.Thread.run(Thread.java:745)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT com.netflix.discovery.shared.transport.TransportException: Cannot execute request on any known server
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.InstanceInfoReplicator.run(InstanceInfoReplicator.java:104)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient.register(DiscoveryClient.java:815)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT com.netflix.discovery.shared.transport.TransportException: Cannot execute request on any known server
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT 2017-03-23 08:43:15.467 WARN [date-service-sa-v1,,,] 19 --- [nfoReplicator-0] c.n.discovery.InstanceInfoReplicator : There was a problem with the instance info replicator
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:111)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.InstanceInfoReplicator.run(InstanceInfoReplicator.java:104)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient.register(DiscoveryClient.java:815)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.lang.Thread.run(Thread.java:745)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$1.execute(EurekaHttpClientDecorator.java:59)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.InstanceInfoReplicator$1.run(InstanceInfoReplicator.java:88)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
Thu Mar 23 2017 10:43:15 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)

```

Then the server is marked as healthy, but it keeps failing to register:

```
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT 2017-03-23 08:48:46.678 ERROR [date-service-sa-v1,,,] 19 --- [tbeatExecutor-0] com.netflix.discovery.DiscoveryClient : DiscoveryClient_date-SERVICE-SA-V1/env1-date-v1-GREEN.apps.cf.site.com:91dfe266-422d-4f5a-69ec-9caf12889a1d - was unable to send heartbeat!
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at java.lang.Thread.run(Thread.java:745)
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at java.lang.Thread.run(Thread.java:745)
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.register(EurekaHttpClientDecorator.java:56)
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient.register(DiscoveryClient.java:815)
Thu Mar 23 2017 10:48:46 GMT+0200 (IST) [APP] OUT at com.netflix.discovery.DiscoveryClient$HeartbeatThread.run(DiscoveryClient.java:1396)


```

Eureka is still pointing to the old instance. There is only one instance registered.

**Comment by csterwa** [comment]
@odedia could you tell us what version of the Spring Cloud Services tile you have installed along with what version your Service Registry service instance is currently at?

**Comment by odedia** [comment]
Tile: Spring Cloud Services: spring-cloud-broker/1.1.1-build.6
Ops Manager: 1.7.16
Not sure what you mean about Service Registry service, I would assume it's the Spring Cloud framework version? We're using Brixton.SR7.

Thanks.

**Comment by csterwa** [comment]
@odedia sorry for taking so long to respond further on this. I am now responding to see if this is still an issue considering you may have upgraded both SCS broker and client connector dependencies? I believe this was resolved in SCS v1.3+ but I'd like to hear if that was the case for you.

**Comment by odedia** [comment]
Sorry, missed this one. I moved on to latest SCS client connector dependencies and don't face issues at the moment.

Thank you!
